0005-ordered-quantification.md

```diff
@@ -0,0 +1,644 @@
+expressions rather than statements allows logic to be used in specifications as well as implementation,
+and reduces the effort Dafny users have to spend on proving their code correct.
+The proposed sequence comprehension expression allows more
+logic that ultimately produces a sequence of values to be expressed as a value rather than a statement.
+Just producing a sequence of the values in the set above, sorted by the natural ordering of `int` values,
+is simple using a sequence comprehension expression: `seq i: int | i in s`.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Both of the proposed new features depend on another new and more fundamental concept: quantification ordering. 
+Therefore, allow me to outline this concept first before then describing `foreach` loops and sequence comprehensions.
+
+## Quantification ordering
+
+Several existing Dafny features support a common syntax for quantifying over one or more variables, internally referred to
+as a "quantifier domain". For example, the `x: nat, y: nat | x < 2 && y < 2` portion of the set comprehension 
+`set x: nat, y: nat | x < 2 && y < 2 :: (x, y)`. The boolean expression after the `|` is referred to as the *range*,
+and I will label the `x: nat` and `y: nat` sections as *quantifier variable declarations*.
+We can augment this existing syntax with a notion of ordering, and allow quantified variables to bind to duplicate values. The key points are:
+
+* Any quantifier domain defines a potentially-infinite, partially-ordered set of *quantifier bindings*.
+* The quantified variable declarations define the values each binding maps to each variable, AND how these bindings are ordered.
+* The range expression after the `|` restricts the quantification to a subset of these bindings, but does not influence their ordering.
+
+A quantifier domain only guarantees an ordering of bindings, 
+but is NOT prescriptive on how to enumerate this domain at runtime, if it is compiled!
+This is consistent with existing expressions such as `set x: real | x in myFiniteSet && P(x)`: 
+although the unfiltered domain of `real` values is not enumerable, the bound `x in myFiniteSet` is, 
+and at runtime this comprehension is calculated by enumerating the contents of `myFiniteSet` 
+and filtering out values that do not satisfy `P(x)`. 
+The new features that are affected by quantification ordering will behave similarly: 
+`seq x: real | x in mySet && P(x)` is calculated via the same filtered enumeration, 
+but collected into a sorted sequence instead.
+
+There will be two supported variable declaration styles, and others could potentially be added in the future:
+
+1. `x [: T]`
+
+    In the existing syntax for quantifier variables (where the type `T` may be explicitly provided or omitted and inferred),
+    the bindings will be ordered according to the *natural ordering* of the type `T`. Not all Dafny types will
```

Comment:
My immediate reaction to the difference between `in` and `<-` is to think of `in` as more "mathematical" and `<-` as more "implementation-focused" (given how those symbols have been used in other contexts before). Because of that, it seems more natural to me to have `<-` impose _more_ ordering, and `in` impose less. What if `in` produced elements in a non-deterministic order, with no duplicates, and `<-` always used a collection-defined ordering with the potential for duplicates. That would eliminate the need for "natural ordering" altogether.

Reply:
Have a look at the latest commit, I went with almost exactly that option. :)

Reply:
🎉